scipy/special/_digamma.pxd

```diff
@@ -0,0 +1,203 @@
+# An implementation of the digamma function for complex arguments.
+#
+# Author: Josh Wilson
+#
+# Distributed under the same license as Scipy.
+#
+# Sources:
+# [1] "The Digital Library of Mathematical Functions", dlmf.nist.gov
+#
+# [2] mpmath (version 0.19), http://mpmath.org
+#
+
+import cython
+from libc.math cimport ceil, fabs, M_PI
+from _complexstuff cimport number_t, nan, zlog, zabs, zdiv
+from _trig cimport sinpi, cospi
+cimport sf_error
+
+cdef extern from "cephes.h":
+    double zeta(double x, double q) nogil
+    double psi(double x) nogil
+
+# Use the asymptotic series for z away from the negative real axis
+# with abs(z) > smallabsz.
+DEF smallabsz = 16
+# Use the reflection principle for z with z.real < 0 that are within
+# smallimag of the negative real axis.
+DEF smallimag = 6
+# Relative tolerance for series
+DEF tol = 2.220446092504131e-16
+# All of the following were computed with mpmath
+# Location of the positive root
+DEF posroot = 1.4616321449683623
+# Value of the positive root
+DEF posrootval = -9.2412655217294275e-17
+# Location of the negative root
+DEF negroot = -0.504083008264455409
+# Value of the negative root
+DEF negrootval = 7.2897639029768949e-17
+
+
+cdef inline double digamma(double z) nogil:
+    """
+    Wrap Cephes' psi to take advantage of the series expansions
+    around the two smallest zeros.
+
+    """
+    if zabs(z - posroot) < 0.5:
+        return zeta_series(z, posroot, posrootval)
+    elif zabs(z - negroot) < 0.3:
+        return zeta_series(z, negroot, negrootval)
+    else:
+        return psi(z)
+
+
+@cython.cdivision(True)
+cdef inline double complex cdigamma(double complex z) nogil:
+    """Compute the digamma function for complex arguments."""
+    cdef:
+        int n
+        double absz = zabs(z)
+        double complex res = 0
+        double complex init
+
+    if z.real <= 0 and ceil(z.real) == z:
+        # Poles
+        sf_error.error("digamma", sf_error.SINGULAR, NULL)
+        return nan + 1j*nan
+    elif zabs(z - negroot) < 0.3:
+        # First negative root
+        return zeta_series(z, negroot, negrootval)
+
+    if z.real < 0 and fabs(z.imag) < smallabsz:
+        # Reflection formula for digamma. See
+        #
+        # http://dlmf.nist.gov/5.5#E4
+        #
+        res -= M_PI*cospi(z)/sinpi(z)
+        z = 1 - z
+        absz = zabs(z)
+
+    if absz < 0.5:
+        # Use one step of the recurrence relation to step away from
+        # the pole.
+        res -= 1/z
+        z += 1
+        absz = zabs(z)
+
+    if zabs(z - posroot) < 0.5:
+        res += zeta_series(z, posroot, posrootval)
+    elif absz > smallabsz:
+        res += asymptotic_series(z)
+    elif z.real >= 0:
+        n = <int>(smallabsz - absz) + 1
```

Comment:
Needs to guard against integer overflow and too many iterations. Define a reasonable maxiter and return nan if too large count required?